<a href="https://colab.research.google.com/github/byrdka/Text-Mining/blob/main/Text_IDF_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

TBANLT 570 - Project 1 by Kevin Byrd

In [1]:
#import libraries
import numpy as np #used to quickly perform mathematical calculations on vectors
import re #regular expressions; used to clean the text data
import sqlite3 #used to interact with the database
import pandas as pd #allows us to work with data using Pandas dataframes
from collections import Counter #used to quickly count letters and words

In [40]:

git init
git add README.md
git commit -m "first commit"
git branch -M main
git remote add origin https://github.com/byrdka/Text-Mining.git
git push -u origin main

SyntaxError: ignored

Load Data

Load Project 01 - Databse.db

In [13]:
#open a connection to the database
conn = sqlite3.connect('Project 01 - Database.db')

#load all documents into a Pandas dataframe named 'df', and use the document_id column as the index
sql = 'SELECT * FROM Article'
df = pd.read_sql_query(sql, conn)

#close database connection
conn.close()

View Database: Contains 3 columns id, category, raw_text. There is 2225 rows.

In [15]:
df

id    category                                           raw_text
0        6347    Politics  Hiding women away in the home hidden behind ve...
1       13840      Sports  Celtic brushed aside Clyde to secure their pla...
2       14775     Unknown  If you have finished Doom 3, Half Life 2 and H...
3       16641     Unknown  Controversial new UK casinos will be banned fr...
4       17511     Unknown  Justine Henin-Hardenne lost to Elena Dementiev...
...       ...         ...                                                ...
2220  9981403  Technology  The US Federal Bureau of Investigation is warn...
2221  9983844      Sports  Scrum-half Matt Dawson is an injury doubt for ...
2222  9986243     Unknown  The US economy has grown more than expected, e...
2223  9987006     Unknown  Three years after a gruelling economic crisis,...
2224  9988130  Technology  BBC Sport unveils its new analysis tool Piero ...

[2225 rows x 3 columns]

In [28]:
#open a connection to the database
conn = sqlite3.connect('Project 01 - Database.db')

#load all documents into a Pandas dataframe named 'df', and use the document_id column as the index
sql = "SELECT * FROM Article WHERE category = 'Entertainment'"
df1 = pd.read_sql_query(sql, conn)

#close database connection
conn.close()

In [29]:
df1

id       category                                           raw_text
0      34150  Entertainment  A saxophone belonging to legendary jazz musici...
1      71079  Entertainment  This year's clutch of Oscar nominees have been...
2      81897  Entertainment  Ocean's Twelve, the crime caper sequel starrin...
3     106635  Entertainment  The late US legend Elvis Presley is likely to ...
4     157219  Entertainment  Comedian Chris Morris, who created controversi...
..       ...            ...                                                ...
192  9331294  Entertainment  The top names in Scottish music are taking par...
193  9471681  Entertainment  US film professionals have declared Fahrenheit...
194  9537397  Entertainment  DJ duo JK and Joel are taking over BBC Radio 1...
195  9663840  Entertainment  Three of the five nominees in the live-action ...
196  9694388  Entertainment  British film director Sir Alan Parker has been...

[197 rows x 3 columns]

Query each category

Business = 270 rows

*   Business = 270 rows
*   Entertainment = 197 rows


*   Politics = 239 rows
*   Sports = 294 rows

*   Technology = 225
*   Unknown = 1000 rows

There are 1225 records with an assigned category.


In [35]:
#define a function that will clean the raw input text in preparation for analysis
def clean_text(raw_text):
  #convert the raw text to lowercase
  text = raw_text.lower()
  #remove all numbers from the text using a regular expression
  text = re.sub(r'[0-9]', ' ', text)
  #remove all underscores from the text
  text = re.sub(r'\_', ' ', text)
  #remove anything else in the text that isn't a word character or a space (e.g., punctuation, special symbols, etc.)
  text = re.sub(r'[^\w\s]', ' ', text)
  #remove any excess whitespace
  for _ in range(10):
    text = text.replace('  ', ' ')
  #remove any leading or trailing space characters
  text = text.strip()
  #return the clean text
  return text

df['clean_text'] = [clean_text(raw_text) for raw_text in df.raw_text] # Clean the raw_text based on the function created above.

df.iloc[0]['clean_text'] # .iloc function enables to select a particular row. This is what returns the clean text.

'celtic brushed aside clyde to secure their place in the scottish cup semi final but only after a nervy and testing first half the home side s craig bryson had a goal chopped off before stan varga headed celtic into the lead alan thompson scored from the penalty spot at the start of the second half after shaun maloney had been fouled stilian petrov slid in a third varga tapped in his second and craig bellamy completed the rout with a fine drive bryn halliwell was the busier keeper early on saving from bellamy chris sutton and juninho clyde had the ball in the net after half an hour through a tremendous strike from bryson but the referee had already blown for a foul by petrov from the resulting free kick darren sheridan curled the ball round the celtic wall only for the post to deny him back at the other end halliwell did well to come off his line and block bellamy s effort to lift the ball over him the keeper misjudged a corner that stephane henchoz headed wide but a similar scenario f

In [37]:
#define a function that will compute the raw letter frequencies for the input texts,
#as well as the total number of letters appearing in the input texts
def letter_counts(input_texts):
  all_text = ' '.join(input_texts) #join all of the input texts into one big string
  letter_counts = Counter(all_text.replace(' ', '')) #count all letters in the text (excluding spaces)
  #return letter counts (sorted from most common to least common),
  #and the total number of letters
  return letter_counts.most_common(), sum(letter_counts.values()) 

df_letter_counts, total_df_letters = letter_counts(df.clean_text)

df_letter_counts

[('e', 51069),
 ('a', 38717),
 ('t', 36947),
 ('n', 31543),
 ('i', 31335),
 ('o', 30158),
 ('s', 27316),
 ('r', 26160),
 ('h', 22100),
 ('l', 19547),
 ('d', 17500),
 ('c', 11972),
 ('u', 11397),
 ('m', 10771),
 ('f', 9452),
 ('g', 9391),
 ('w', 9029),
 ('p', 8401),
 ('b', 7644),
 ('y', 7601),
 ('v', 4554),
 ('k', 3982),
 ('j', 1237),
 ('x', 658),
 ('z', 445),
 ('q', 392)]